hrm-service/src/contact-job/contact-job-complete.ts

```diff
@@ -73,6 +75,76 @@ const processCompletedContactJob = async (completedJob: CompletedContactJobBody)
   }
 };
 
+export const getAttemptNumber = async (completedJob: CompletedContactJobBody, contactJob: ContactJobRecord) => {
+  // Attempt number can be null when error occurs in lambda before job data is parsed
+  // In this case we need to fetch the job from the database to get the attempt number
+  if (completedJob.attemptNumber != null) {
+    return completedJob.attemptNumber;
+  }
+
+  return  contactJob.numberOfAttempts;
+};
+
+export const getContactJobOrFail = async (completedJob: CompletedContactJobBody) => {
+  const contactJob = await getContactJobById(completedJob.jobId);
+
+  if (!contactJob) {
+    throw new Error(`Could not find contact job with id ${completedJob.jobId}`);
+  }
+
+  return contactJob;
+};
+
+export const handleSuccess = async (completedJob: CompletedContactJobBody) => {
+  if (completedJob.attemptResult !== 'success') return;
+
+  await processCompletedContactJob(completedJob);
+
+  // Mark the job as completed
+  const completionPayload = {
+    message: 'Job processed successfully',
+    value: completedJob.attemptPayload,
+  };
+  const markedComplete = await completeContactJob(completedJob.jobId, completionPayload);
+
+  return markedComplete;
+};
+
+export const handleFailure = async (completedJob: CompletedContactJobBody, jobMaxAttempts: number) => {
+  if (completedJob.attemptResult === 'success') return;
+
+  const { jobId } = completedJob;
+  let { attemptPayload } = completedJob;
+
+  if (typeof attemptPayload !== 'string') {
+    attemptPayload = "Message did not contain attemptPayload. Likely DLQ'd from lambda";
+  }
+
+  // emit an error to pick up in metrics since completed queue is our
+  // DLQ. These may be duplicates of ContactJobProcessorErrors that have
+  // already caused an alarm, but there is a chance of other errors ending up here.
+  console.error(
+    new ContactJobCompleteProcessorError(
+      `process job with id ${jobId} failed`,
+      attemptPayload,
+    ),
+  );
+
+  const contactJob = await getContactJobOrFail(completedJob);
+  const attemptNumber = await getAttemptNumber(completedJob, contactJob);
```

Comment:
Omit me if I'm wrong, but I believe the call to `getContactJobOrFail` can be moved within `getAttemptNumber`, only if `completedJob.attemptNumber` is missing, and avoid that one extra call in most cases?

Reply:
I had it that way beforehand,  but then realized that if there is no contactJob for some reason, we really shouldn't be doing anything else and should fail immediately because nothing else to do. Really `getContactJobOrFail` should immediately end retries instead of just throwing a regular exception, I think.